src/core/lib/surface/channel_init.cc

```diff
@@ -20,34 +20,323 @@
+    //  - If B is registered as BeforeAll, then A will be 'after' B here.
+    for (const auto& filter : filters) {
+      auto dep_it = original.find(filter.filter);
+      std::string after_str;
+      if (dep_it != original.end() && !dep_it->second.empty()) {
+        after_str = absl::StrCat(
+            std::string(max_filter_name_len + 1 -
+                            strlen(NameFromChannelFilter(filter.filter)),
+                        ' '),
+            "after ",
+            absl::StrJoin(
+                dep_it->second, ", ",
+                [](std::string* out, const grpc_channel_filter* filter) {
+                  out->append(NameFromChannelFilter(filter));
+                }));
+      }
+      const auto filter_str =
+          absl::StrCat("  ", loc_strs[filter.filter],
+                       NameFromChannelFilter(filter.filter), after_str);
+      gpr_log(GPR_INFO, "%s", filter_str.c_str());
+    }
+    // Finally list out the terminal filters and where they were registered
+    // from.
+    for (const auto& terminal : terminal_filters) {
+      const auto filter_str = absl::StrCat(
+          "  ", loc_strs[terminal.filter],
+          NameFromChannelFilter(terminal.filter),
+          std::string(max_filter_name_len + 1 -
+                          strlen(NameFromChannelFilter(terminal.filter)),
+                      ' '),
+          "[terminal]");
+      gpr_log(GPR_INFO, "%s", filter_str.c_str());
+    }
+  }
+  return StackConfig{std::move(filters), std::move(terminal_filters),
+                     std::move(post_processor_functions)};
+};
+
 ChannelInit ChannelInit::Builder::Build() {
   ChannelInit result;
   for (int i = 0; i < GRPC_NUM_CHANNEL_STACK_TYPES; i++) {
-    auto& slots = slots_[i];
-    std::stable_sort(
-        slots.begin(), slots.end(),
-        [](const Slot& a, const Slot& b) { return a.priority < b.priority; });
-    auto& result_slots = result.slots_[i];
-    result_slots.reserve(slots.size());
-    for (auto& slot : slots) {
-      result_slots.emplace_back(std::move(slot.stage));
-    }
+    result.stack_configs_[i] =
+        BuildStackConfig(filters_[i], post_processors_[i],
+                         static_cast<grpc_channel_stack_type>(i));
   }
   return result;
 }
 
 bool ChannelInit::CreateStack(ChannelStackBuilder* builder) const {
-  for (const auto& stage : slots_[builder->channel_stack_type()]) {
-    if (!stage(builder)) return false;
+  const auto& stack_config = stack_configs_[builder->channel_stack_type()];
+  for (const auto& filter : stack_config.filters) {
+    if (!filter.predicate(builder->channel_args())) continue;
+    builder->AppendFilter(filter.filter);
+  }
+  bool found_terminator = false;
+  for (const auto& terminator : stack_config.terminators) {
+    if (!terminator.predicate(builder->channel_args())) continue;
+    builder->AppendFilter(terminator.filter);
+    found_terminator = true;
+    break;
```

Comment:
Instead of stopping here, how about continuing and generating an error if any of the other terminal filters' predicates say they should be added?  It seems like that should be an error rather than just silently taking the first one in sorted order by name.

Reply:
Terminators are collected in registration order (I'm not sure that we sort them) - and in any case the one example we have of this everything is currently called `connected_channel`. I'd be ok with issuing a warning here if it fails, but I think we're in 'safe to recover' territory here if two match so I'd hesitate to do something more severe.

Reply:
It's not a name sort. I'm a bit on the fence about this because it's a recoverable error kind of, but let's try it and find out some. Made it an error for multiples (or none) to match.